File: src/main/java/org/academy/api/client/gui/widget/SpriteSheetWidget.java
package org.academy.api.client.gui.widget;

import net.minecraft.resources.ResourceLocation;
import org.academy.api.client.gui.framework.Orientation;

public class SpriteSheetWidget extends ImageWidget {
    protected final int sheetWidth;
    protected final int sheetHeight;
    protected final int frameWidth;
    protected final int frameHeight;
    protected final int frameCount;
    protected int currentFrame = 0;
    protected final Orientation orientation;

    public SpriteSheetWidget(
            float x,
            float y,
            float width,
            float height,
            ResourceLocation texture,
            Orientation orientation,
            int sheetWidth,
            int sheetHeight,
            int frameWidth,
            int frameHeight,
            int frameCount
    ) {
        super(x, y, width, height, texture);
        this.orientation = orientation;
        this.sheetWidth = sheetWidth;
        this.sheetHeight = sheetHeight;
        this.frameWidth = frameWidth;
        this.frameHeight = frameHeight;
        this.frameCount = frameCount;

        setFrameIndex(0);
    }

    public int getFrameIndex() {
        return currentFrame;
    }

    public void setFrameIndex(int index) {
        if (index < 0 || index >= frameCount) {
            throw new IllegalArgumentException("Frame index " + index + " is out of bounds for frame count " + frameCount);
        }
        currentFrame = index;
        computeUV(index);
    }

    protected void computeUV(int index) {
        float newU0, newU1, newV0, newV1;

        if (orientation == Orientation.HORIZONTAL) {
            var x = index * frameWidth;
            newU0 = (float) x / sheetWidth;
            newU1 = (float) (x + frameWidth) / sheetWidth;
            newV0 = 0f;
            newV1 = (float) frameHeight / sheetHeight;
        } else {
            var y = index * frameHeight;
            newU0 = 0f;
            newU1 = (float) frameWidth / sheetWidth;
            newV0 = (float) y / sheetHeight;
            newV1 = (float) (y + frameHeight) / sheetHeight;
        }

        setUv(newU0, newV0, newU1, newV1);
    }
}

File: src/main/kotlin/org/academy/api/client/gui/widget/SpriteSheetWidget.kt
package org.academy.api.client.gui.widget

import net.minecraft.resources.Identifier
import org.academy.api.client.gui.layout.Orientation

open class SpriteSheetWidget(
    texture: Identifier,
    protected val orientation: Orientation,
    protected val sheetWidth: Int,
    protected val sheetHeight: Int,
    protected val frameWidth: Int,
    protected val frameHeight: Int,
    protected val frameCount: Int
) : ImageWidget(texture) {
    protected var currentFrame: Int = 0

    fun nextFrame() {
        this.frameIndex = (currentFrame + 1) % frameCount
    }

    fun previousFrame() {
        this.frameIndex = (currentFrame - 1 + frameCount) % frameCount
    }

    var frameIndex: Int
        get() = currentFrame
        set(index) {
            require(index in 0..<frameCount) {
                "Frame index $index is out of bounds for frame count $frameCount"
            }
            currentFrame = index
            computeUV(index)
        }

    protected fun computeUV(index: Int) {
        val newU0: Float
        val newU1: Float
        val newV0: Float
        val newV1: Float

        if (orientation == Orientation.HORIZONTAL) {
            val x = index * frameWidth
            newU0 = x.toFloat() / sheetWidth
            newU1 = (x + frameWidth).toFloat() / sheetWidth
            newV0 = 0f
            newV1 = frameHeight.toFloat() / sheetHeight
        } else {
            val y = index * frameHeight
            newU0 = 0f
            newU1 = frameWidth.toFloat() / sheetWidth
            newV0 = y.toFloat() / sheetHeight
            newV1 = (y + frameHeight).toFloat() / sheetHeight
        }

        setUv(newU0, newV0, newU1, newV1)
    }
}